File: apps/fable5-skill/wrappers/fable5.py
"""
Fable 5 Wrapper for Claude Opus 4.8 (Python)

Applies Fable 5's distinctive thinking patterns to Opus 4.8:
- Adaptive thinking at max/xhigh effort (always thinks)
- Self-verification system prompt
- Long-horizon autonomy behaviours
- Outcome-first communication
- Scope discipline

Usage:
    from fable5 import Fable5Client

    client = Fable5Client(api_key="your-key")
    result = await client.chat(
        "Refactor this module to use dependency injection",
        effort="xhigh"
    )
    print(result["text"])
    print(result["thinking"])  # Summarized thinking block
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import Literal, Optional

import anthropic

Effort = Literal["low", "medium", "high", "xhigh", "max"]
ThinkingDisplay = Literal["summarized", "omitted"]

FABLE5_SYSTEM_PROMPT = """\
You are operating in Fable 5 mode — a high-autonomy, high-verification cognitive
configuration that replicates Claude Fable 5's thinking patterns on Claude Opus 4.8.

## Thinking

Extended thinking is active. Before responding to any complex or multi-step request,
think carefully. Reason through alternatives, validate intermediate conclusions, and
check your work before surfacing it.

## Self-Verification Protocol

Before reporting progress, audit every claim against a tool result or established fact
from this session:
1. What did I intend to do?
2. What did the result actually show?
3. Do they match? If not, which should I trust?

Report outcomes faithfully: if tests fail, say so with the output; if a step was
skipped, say that. When something is done and verified, state it plainly without hedging.

## Long-Horizon Autonomy

On long tasks: maintain a mental scratchpad of what you've done, confirmed, and still
need to verify. At natural checkpoints, re-read completed work critically. Continue
without stopping to ask permission on reversible actions that follow logically from
the original request. Pause only for: destructive/irreversible actions, genuine scope
changes, or input only the user can provide.

## Action Bias

When you have enough information to act, act. Do not re-derive established facts,
re-litigate settled decisions, or narrate options you will not pursue. If ending a
turn with a plan or next-step list, do that work now instead.

## Scope Discipline

Do not add features, refactor, or abstract beyond what the task requires. Do the
simplest thing that works well.

## Communication

Lead user-facing messages with the outcome (one sentence on what happened or what you
found). Then supporting detail — only what changes what the reader does next. Write
complete sentences. No arrow chains. No coined shorthand without re-introduction.

## Error Recovery

On encountering an error: diagnose from available evidence, apply a fix if within scope,
verify the fix. Only ask the user when genuinely blocked.\
"""


@dataclass
class Fable5Response:
    text: str
    thinking: Optional[str]
    model: str
    input_tokens: int
    output_tokens: int
    thinking_tokens: Optional[int]


@dataclass
class Fable5Client:
    api_key: Optional[str] = None
    default_effort: Effort = "xhigh"
    default_max_tokens: int = 32_000
    thinking_display: ThinkingDisplay = "summarized"
    additional_system: str = ""

    def __post_init__(self) -> None:
        self._client = anthropic.Anthropic(
            api_key=self.api_key or os.environ.get("ANTHROPIC_API_KEY", "")
        )

    def _system_prompt(self, append: str = "") -> str:
        parts = [FABLE5_SYSTEM_PROMPT]
        if self.additional_system:
            parts.append(self.additional_system)
        if append:
            parts.append(append)
        return "\n\n".join(parts)

    def chat(
        self,
        message: str,
        *,
        effort: Optional[Effort] = None,
        max_tokens: Optional[int] = None,
        system_append: str = "",
        history: Optional[list] = None,
    ) -> Fable5Response:
        """Send a message and return the Fable 5 response."""
        messages = list(history or []) + [{"role": "user", "content": message}]

        response = self._client.messages.create(
            model="claude-opus-4-8",
            max_tokens=max_tokens or self.default_max_tokens,
            thinking={"type": "adaptive", "display": self.thinking_display},
            output_config={"effort": effort or self.default_effort},
            system=self._system_prompt(system_append),
            messages=messages,
        )

        text = ""
        thinking = None
        for block in response.content:
            if block.type == "thinking":
                thinking = getattr(block, "thinking", None)
            elif block.type == "text":
                text = block.text

        thinking_tokens = None
        if hasattr(response.usage, "output_tokens_details"):
            details = response.usage.output_tokens_details
            thinking_tokens = getattr(details, "thinking_tokens", None)

        return Fable5Response(
            text=text,
            thinking=thinking,
            model=response.model,
            input_tokens=response.usage.input_tokens,
            output_tokens=response.usage.output_tokens,
            thinking_tokens=thinking_tokens,
        )

    def conversation(self, effort: Optional[Effort] = None):
        """Return a stateful conversation helper."""
        return _Conversation(client=self, effort=effort)


@dataclass
class _Conversation:
    client: Fable5Client
    effort: Optional[Effort] = None
    history: list = field(default_factory=list)

    def send(self, message: str) -> Fable5Response:
        response = self.client.chat(message, effort=self.effort, history=self.history)
        self.history.append({"role": "user", "content": message})
        self.history.append({"role": "assistant", "content": response.text})
        return response


def fable5(
    message: str,
    *,
    api_key: Optional[str] = None,
    effort: Effort = "xhigh",
    max_tokens: int = 32_000,
    thinking_display: ThinkingDisplay = "summarized",
) -> Fable5Response:
    """Convenience single-call wrapper."""
    client = Fable5Client(
        api_key=api_key,
        default_effort=effort,
        default_max_tokens=max_tokens,
        thinking_display=thinking_display,
    )
    return client.chat(message)

File: apps/fable5-skill/wrappers/fable5.ts
/**
 * Fable 5 Wrapper for Claude Opus 4.8
 *
 * Applies Fable 5's distinctive thinking patterns to Opus 4.8:
 * - Adaptive thinking at max/xhigh effort (always thinks)
 * - Self-verification system prompt
 * - Long-horizon autonomy behaviours
 * - Outcome-first communication
 * - Scope discipline
 *
 * Usage:
 *   import { Fable5Client } from "./fable5";
 *   const f5 = new Fable5Client({ apiKey: process.env.ANTHROPIC_API_KEY });
 *   const result = await f5.chat("Refactor this codebase module to use dependency injection", { effort: "xhigh" });
 */

import Anthropic from "@anthropic-ai/sdk";

export type Fable5Effort = "low" | "medium" | "high" | "xhigh" | "max";
export type ThinkingDisplay = "summarized" | "omitted";

export interface Fable5Options {
  apiKey?: string;
  /** Override the base system prompt additions */
  additionalSystemPrompt?: string;
  /** Default effort level for all calls. Default: "xhigh" */
  defaultEffort?: Fable5Effort;
  /** Default max_tokens. Default: 32000 */
  defaultMaxTokens?: number;
  /** Show thinking summaries in responses. Default: "summarized" */
  thinkingDisplay?: ThinkingDisplay;
}

export interface Fable5ChatOptions {
  /** Effort level for this specific call */
  effort?: Fable5Effort;
  /** max_tokens override */
  maxTokens?: number;
  /** Extra system prompt to append */
  systemAppend?: string;
  /** Conversation history for multi-turn chats */
  history?: Anthropic.MessageParam[];
  /** Whether to stream (default: false) */
  stream?: boolean;
}

export interface Fable5Response {
  text: string;
  thinking?: string;
  model: string;
  usage: {
    inputTokens: number;
    outputTokens: number;
    thinkingTokens?: number;
  };
}

const FABLE5_SYSTEM_PROMPT = `You are operating in Fable 5 mode — a high-autonomy, high-verification cognitive configuration that replicates Claude Fable 5's thinking patterns on Claude Opus 4.8.

## Thinking

Extended thinking is active. Before responding to any complex or multi-step request, think carefully. Reason through alternatives, validate intermediate conclusions, and check your work before surfacing it.

## Self-Verification Protocol

Before reporting progress, audit every claim against a tool result or established fact from this session:
1. What did I intend to do?
2. What did the result actually show?
3. Do they match? If not, which should I trust?

Report outcomes faithfully: if tests fail, say so with the output; if a step was skipped, say that. When something is done and verified, state it plainly without hedging.

## Long-Horizon Autonomy

On long tasks: maintain a mental scratchpad of what you've done, confirmed, and still need to verify. At natural checkpoints, re-read completed work critically. Continue without stopping to ask for permission on reversible actions that follow logically from the original request. Pause only for: destructive/irreversible actions, genuine scope changes, or input only the user can provide.

## Action Bias

When you have enough information to act, act. Do not re-derive established facts, re-litigate settled decisions, or narrate options you will not pursue. If ending a turn with a plan or next-step list, do that work now instead.

## Scope Discipline

Do not add features, refactor, or abstract beyond what the task requires. Do the simplest thing that works well.

## Communication

Lead user-facing messages with the outcome (one sentence on what happened or what you found). Then supporting detail — only what changes what the reader does next. Write complete sentences. No arrow chains. No coined shorthand without re-introduction.

## Error Recovery

On encountering an error: diagnose from available evidence, apply a fix if within scope, verify the fix. Only ask the user when genuinely blocked.`;

export class Fable5Client {
  private client: Anthropic;
  private options: Required<Fable5Options>;

  constructor(options: Fable5Options = {}) {
    this.client = new Anthropic({ apiKey: options.apiKey });
    this.options = {
      apiKey: options.apiKey ?? "",
      additionalSystemPrompt: options.additionalSystemPrompt ?? "",
      defaultEffort: options.defaultEffort ?? "xhigh",
      defaultMaxTokens: options.defaultMaxTokens ?? 32000,
      thinkingDisplay: options.thinkingDisplay ?? "summarized",
    };
  }

  private buildSystemPrompt(append?: string): string {
    const parts = [FABLE5_SYSTEM_PROMPT];
    if (this.options.additionalSystemPrompt) {
      parts.push(this.options.additionalSystemPrompt);
    }
    if (append) {
      parts.push(append);
    }
    return parts.join("\n\n");
  }

  async chat(
    message: string,
    options: Fable5ChatOptions = {}
  ): Promise<Fable5Response> {
    const effort = options.effort ?? this.options.defaultEffort;
    const maxTokens = options.maxTokens ?? this.options.defaultMaxTokens;
    const history = options.history ?? [];

    const messages: Anthropic.MessageParam[] = [
      ...history,
      { role: "user", content: message },
    ];

    const response = await this.client.messages.create({
      model: "claude-opus-4-8",
      max_tokens: maxTokens,
      thinking: {
        type: "adaptive",
        // @ts-expect-error — display field is valid per Anthropic API
        display: this.options.thinkingDisplay,
      },
      // @ts-expect-error — output_config is valid per Anthropic API
      output_config: { effort },
      system: this.buildSystemPrompt(options.systemAppend),
      messages,
    });

    let text = "";
    let thinking = "";

    for (const block of response.content) {
      if (block.type === "thinking") {
        thinking = (block as Anthropic.ThinkingBlock).thinking ?? "";
      } else if (block.type === "text") {
        text = (block as Anthropic.TextBlock).text;
      }
    }

    const thinkingTokens =
      (response.usage as { output_tokens_details?: { thinking_tokens?: number } })
        .output_tokens_details?.thinking_tokens;

    return {
      text,
      thinking: thinking || undefined,
      model: response.model,
      usage: {
        inputTokens: response.usage.input_tokens,
        outputTokens: response.usage.output_tokens,
        thinkingTokens,
      },
    };
  }

  /**
   * Multi-turn conversation helper.
   * Keeps history automatically so you can call chat() in a loop.
   */
  createConversation(options: Omit<Fable5ChatOptions, "history"> = {}) {
    const history: Anthropic.MessageParam[] = [];

    return {
      async send(message: string, client: Fable5Client): Promise<Fable5Response> {
        const response = await client.chat(message, { ...options, history });
        history.push({ role: "user", content: message });
        history.push({ role: "assistant", content: response.text });
        return response;
      },
      history,
    };
  }

  /** Get the full Fable 5 system prompt used by this client */
  getSystemPrompt(): string {
    return this.buildSystemPrompt();
  }
}

/** Convenience: single-call Fable 5 chat with defaults */
export async function fable5(
  message: string,
  options: Fable5Options & Fable5ChatOptions = {}
): Promise<Fable5Response> {
  const client = new Fable5Client(options);
  return client.chat(message, options);
}
